source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+/*******************************************************************************
+
+    Defines a monetary type used in the blockchain
+
+   This struct can hold an amount between 0 and 500_000_000 * 10^7,
+   as there is an absolute maximum supply of 500M and each coin can be divided
+   in up to 10^7 units.
+
+   This struct does not expose operation overloading on purpose.
+   Having operator overloading would mean that any error should be reported
+   as either an `assert` being triggered or an `Exception` being thrown.
+   The convenience of using `a + b` would likely mean some places wouldn't
+   be properly bound checked, which would in turn open a DoS.
+   Instead, we provide two kind of functions for operations:
+   `bool OPNAME(ref Amount res, const Type)` and `Amount mustOPNAME(const Type)`
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Amount;
+
+/// Ditto
+public struct Amount
+{
+    import agora.common.Deserializer;
+    import agora.common.Hash;
+    import agora.common.Serializer;
+
+    import core.checkedint;
+
+    /// Number of units ('cents') per coins
+    public static immutable Amount UnitPerCoin   = Amount(10_000_000, true);
+    /// Maximum number of BOA coins that can ever be in circulation
+    public static immutable ulong  MaxCoinSupply = 500_000_000;
+    /// Maximum amount of money that can ever be in circulation
+    public static immutable Amount MaxUnitSupply =
+        Amount(UnitPerCoin.value * MaxCoinSupply, true);
+
+    /// Internal data storage
+    private ulong value;
+
+    /// Support for hashing
+    public void computeHash(scope HashDg dg) const nothrow @nogc @safe
+    {
+        hashPart(this.value, dg);
+    }
+
+    /// Support for serialization
+    public void serialize(scope SerializeDg dg) const nothrow @safe
+    {
+        serializePart(this.value, dg);
+    }
+
+    /// Support for deserialization
+    public void deserialize(scope DeserializeDg dg) nothrow @safe
+    {
+        deserializePart(this.value, dg);
+    }
+
+    /// Pretty-print this value
+    public void toString (SinkT dg) const @safe
+    {
+        import std.format;
+        formattedWrite(dg, "%d", this.value);
+    }
+
+    /// Also support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        string ret;
+        scope SinkT dg = (scope v) { ret ~= v; };
+        this.toString(dg);
+        return ret;
+    }
+
+    private alias SinkT = void delegate(scope const(char)[] v) @safe;
+
+    /// Support for Vibe.d deserialization
+    public static Amount fromString (scope const(char)[] str) pure @safe
+    {
+        import std.conv : to;
+        immutable ul = str.to!ulong;
+        if (!Amount.isInRange(ul))
```

Comment:
So in some cases we poison the value, in other cases we throw an Exception. Is there a general rule where one or the other should be preferred?

Reply:
Also I feel this check should be in the constructor?

Reply:
This is a bit of an odd one: `fromString` is required for Vibe.d. But actually, Vibe.d has a serialization framework which offers multiple alternatives. I'd like to unify everything under this architecture ASAP, because every new type we introduce will need to define 5 methods...

Reply:
And the check is not in the constructor because the constructor being called with wrong parameter is a programming bug. While `fromString` is called for deserialization of parameters sent to the `API`.